kedro/contrib/config/templated_config.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2018-2019 QuantumBlack Visual Analytics Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND
+# NONINFRINGEMENT. IN NO EVENT WILL THE LICENSOR OR OTHER CONTRIBUTORS
+# BE LIABLE FOR ANY CLAIM, DAMAGES, OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF, OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# The QuantumBlack Visual Analytics Limited ("QuantumBlack") name and logo
+# (either separately or in combination, "QuantumBlack Trademarks") are
+# trademarks of QuantumBlack. The License does not grant you any right or
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+#     or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module provides ``kedro.config`` with the functionality to load one
+or more configuration files from specified paths, and replace template strings with default values.
+"""
+import re
+from typing import Any, Dict, List, Optional, Union
+
+import jmespath
+
+from kedro.config import ConfigLoader
+
+
+class TemplatedConfigLoader(ConfigLoader):
+    """
+    Extension of the ConfigLoader class that allows for template values, wrapped in brackets like:
+     ${..}, to be replaced by default values.
+
+    Default values are provided in a dictionary.
+    """
+
+    def resolve(self, patterns: Union[str, List[str]],
```

Comment:
I think the first option gives some nice properties, as you are forcing the user to explicitly define the `arg_values` argument, and if they don't it will just behave as the parent `get` method.

The second one is problematic because you want the `patterns` and `arg_values` argument are inherently tied together. Put differently, you want to allow the user to specify different `arg_values` arguments for different `patterns`.